packages/realm/src/schema/property_schema_grammar.txt

```diff
@@ -0,0 +1,114 @@
+---------------------------------------------------------------------
+
+
+----------------
+GRAMMAR NOTATION
+----------------
+
+<rule name>:<rule>      Rule structure
+<single quoted string>  Exact literal
+<double quoted string>  Exact string literal
+<uppercase word>        Varying literal
+<lowercase word>        A call to another rule
+
+:                       Start of rule 
+;                       End of rule
+x*                      Match zero or more of x
+x+                      Match one or more of x
+x?                      Match zero or one of x
+x1 x2                   Match x1 then x2
+x1 | x2                 Match x1 or x2
+( x )                   Match x (everything inside the parentheses)
+< ... >                 Description (...) of a rule
+//                      Comment with no semantic meaning
+
+
+----------------
+START RULE
+----------------
+
+propertySchema:
+    propertySchemaString
+    | propertySchemaObject
+    ;
+
+
+----------------
+OTHER RULES
+----------------
+
+primitiveType:
+    "bool"
+    | "int"
+    | "float"
+    | "double"
+    | "decimal128"
+    | "objectId"
+    | "string"
+    | "data"
+    | "date"
+    | "mixed"
+    | "uuid"
+    ;
+
+collectionType:
+    "list"
+    | "dictionary"
+    | "set"
+    ;
+
+relationshipType:
+    "object"
+    | "linkingObjects"
+    ;
+
+propertySchemaString:
+    ( primitiveType | USER_TEXT ) ( "?" )? ( "[]" | "{}" | "<>" )?
+    ;
+
+propertySchemaObject:
+    '{' propertySchemaObjectFields '}'
+    ;
+
+// Only showing the fields 'type', 'objectType', 'optional', and 'property'
+propertySchemaObjectFields:
+    'type' ':' ( primitiveType | collectionType | relationshipType ) ','
+    'objectType' ':' ( primitiveType | USER_TEXT | 'undefined' ) ','
+    'property' ':' ( USER_TEXT | 'undefined' ) ','
+    'optional' ':' ( 'true' | 'false' | 'undefined' ) ','
+    ;
+
+// Logic-based propertySchemaObjectFields
+propertySchemaObjectFields:
+    'type' ':' ( primitiveType | collectionType | relationshipType ) ','
+
+    // For objectType:
+    // if type is primitiveType
+        'objectType' ':' 'undefined' ','
+    // else if type is collectionType
+        'objectType' ':' ( primitiveType | USER_TEXT ) ','
+    // else (relationshipType)
+        'objectType' ':' USER_TEXT ','
+
+    // For property:
+    // if type is "linkingObjects"
+        'property' ':' USER_TEXT ','
+    // else
+        'property' ':' 'undefined' ','
+
+    // For optional:
+    // if type is "mixed" or objectType is "mixed" or type is "object" or (type is "dictionary" and objectType is USER_TEXT)
+        'optional' ':' ( 'true' | 'undefined' ) ','
+    // else if objectType is USER_TEXT and type is "list" or "set" or "linkingObjects"
+        'optional' ':' ( 'false' | 'undefined' ) ','
+    // else
+        'optional' ':' ( 'true' | 'false' | 'undefined' ) ','
+    ;
+
+USER_TEXT:
+    "<any Unicode character>+"   // Excluding reserved type names
```

Comment:
Went a little loose on that one. Should probably be `'"' <any Unicode character>+ '"'` (only showing use of double quotes).

Since I initially made the grammar only for us to use a quick overview I "friendlified" a bit. For instance, that's why I made `"` represent a string literal (rather than showing that both double and single quotes can be used in JS when representing strings).

But I'll clean the grammar up and make it a bit stricter since we want to keep it as docs 👍

Reply:
As I mentioned, I don't think we need to deal with the difference between ' and " (or even \`) strings. I think once you TS-ify the object portion of the grammar, it will be a lot simpler. 

In this case though, I'd just use a prose description like "Any non-empty string that doesn't match a reserved type name" (without the quotes). Since this is intended for human readers rather than a parser generator, we can be a bit looser on the formality for things like this.

Reply:
Also, consider naming this rule `userType` or `UserType` rather than `USER_TEXT`. I think it is best to emphasize that semantically it is the name of a type, not just arbitrary text (even if lexically it is - we don't stop someone from writing a paragraph as their type name, even if it is a terrible idea)